site-src/geps/gep-851.md

```diff
@@ -0,0 +1,102 @@
+# GEP-851: Allow Multiple Certificate Refs per Gateway Listener
+
+* Issue: [#851](https://github.com/kubernetes-sigs/gateway-api/issues/851)
+* Status: Implementable
+
+## TLDR
+
+Replace `CertificateRef` field with an `CertificateRefs` field in Gateway
+Listeners.
+
+## Goals
+
+* Allow certs to be attached for multiple unique Hostnames for the same
+  Listener. This is most helpful in the following situations:
+  * The hostname on the Listener is empty or unspecified allowing Routes to
+    attach arbitrary hostnames to the Listener with Routes.
+  * The hostname on the Listener is a wildcard and Route owners want to attach
+    multi-level subdomains. For example, `*.example.com` on the Listener and
+    `app.dev.example.com` on a Route.
+* Enable specifying RSA and ECDSA certs on the same Listener.
+* Improve cert upgrades by allowing new and old certs to be attached to the same
+  Listener.
+
+## Non-Goals
+
+* Be too prescriptive about how multiple certificates are handled for a Gateway
+  Listener. Each implementation may handle this slightly differently.
+
+## Introduction
+
+When we removed certificate references from HTTPRoute ([GEP
+746](/geps/gep-746/)) we lost the capability to attach multiple certificates to
+a Listener. This GEP proposes adding that capability back, directly to the
+Gateway Listener this time.
+
+In addition to helping with the self service approach, where app owners may
+attach multiple hostnames (and certs) to a single Listener, this can help with
+other use cases:
+
+* Attaching RSA and ECDSA certs.
+* Attaching both new and old certs during an upgrade.
+
+## API
+
+The `CertificateRef` field in `GatewayTLSConfig` would be replaced with the
+following `CertificateRefs` field:
+
+```go
+    // CertificateRefs contains a series of references to Kubernetes objects that
+    // contains TLS certificates and private keys. These certificates are used to
+    // establish a TLS handshake for requests that match the hostname of the
+    // associated listener.
+    //
+    // When multiple TLS certificates are specified, SNI matching MUST be used
+    // to determine the appropriate certificate to use.
```

Comment:
Does the API provide any guidance on how to pick a certificate when multiple TLS certs can be used for a given TLS handshake context (context = SNI+cipher+othe TLS params)?

Reply:
I'm not sure if there's any guidance we can provide that could be universally implemented here. I think at least in some cases this behavior is going to be pretty locked into the implementation without a lot of room for customization.

Reply:
In that case, we have two options:
1. no mention of this
2. mention "Such situations are possible and behaviors will be implementation-specific"

I don't have a preference. You pick one and consider this thread resolved.

Reply:
I tried 2 first, but every combination of wording I could come up with felt more confusing to me than not saying anything. Open to any suggestions though.

Reply:
I think we need to define the hostname matching rules like we do with the Listener/Route ones, something like:

"If there is an overlap between certificate SNI hostnames, then any precise matches take precedence. When two precise matches are relevant, the longest hostname wins. For multiple matches outside of the cases, the implementation decides".